Add Dir::is_dir and refactor Dir::create_dir_all

kubkon opened the pull request:
This commit adds `Dir::is_dir` and refactors `Dir::create_dir_all`
which makes use of it internally. This commit also enables several
tests testing `create_dir_all` along both positive and negative
execution paths.

sunfishcode commented on tests/fs.rs:
```diff
@@ -535,18 +535,18 @@ fn mkdir_path_already_exists_error() {
     let tmpdir = tmpdir();
     let dir = "d1/d2";
     check!(tmpdir.create_dir_all(dir));
-    assert!(dir.is_dir())
```
This should translate into `assert!(tmpdir.is_dir(dir));`

sunfishcode commented on tests/fs.rs:
```diff
@@ -535,18 +535,18 @@ fn mkdir_path_already_exists_error() {
+    assert!(tmpdir.is_dir("d1"));
+    let dir = check!(tmpdir.open_dir("d1"));
+    assert!(dir.is_dir("d2"));
+    assert!(tmpdir.is_dir("d1/d2"));
```
These extra checks are good, though would it be inconvenient to add them in a separate file? That way we can keep `fs.rs` as close to a direct translation of Rust's `fs.rs` as possible.

kubkon commented:
OK, @sunfishcode, I've now addressed your comments in e5a077a. Have a look and lemme know what you think. I've had to do some small refactoring to get `check!` and other macros available across the entire `tests` crate.

sunfishcode commented:
Looks great!
